### PR TITLE
Issue #545 - PlansAllowedInBundle

### PR DESCRIPTION
**Pull request by javier-gomez-hs**
- Added new control of the attribute plansAllowedInBundle, to limit the number of plans allowed per bundle (in progress)

### CODE REVIEW

File: subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java [hunk 1]
```diff
@@ -221,6 +244,18 @@ public SubscriptionBase createSubscription(final UUID bundleId, final PlanPhaseS
                     }
                 }
 
+                // verify the number of subscriptions (of the same kind) allowed per bundle and the existing ones
+                if (ProductCategory.ADD_ON.toString().equalsIgnoreCase(plan.getProduct().getCategory().toString())) {
+                    if (plan.getPlansAllowedInBundle() != -1 && plan.getPlansAllowedInBundle() > 0) {
+                        int existingAddOnsWithSamePlanName = countExistingAddOnsWithSamePlanName(getSubscriptionsForBundle(bundleId, null, context), plan.getName());
```
Comment (sbrossie) [inline subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java]: Make the `getSubscriptionsForBundle` outside of the `for` loop as it is quite expensive and should only needed once.

**Comment by sbrossie** [comment]
Overall logic + tests look good.

However:
* I am not seeing the logic (not the test) for the  `changePlan` use case?
* I had one comment about performance in the `createBaseSubscriptionWithAddOns` with regard to `getSubscriptionForBundle`.
* Optional (but something to think of): Regarding use of `getSubscriptionForBundle`, I am wondering if there is a way to avoid making that call and use a less expensive DAO call to just retrieve what we need (or maybe additional info when retrieving the bundle

**Comment by javier-gomez-hs** [comment]
@sbrossie 
Yes, you are right I forgot to commit the `changePlan` logic / test.
I'll work on that and the possibility to replace the `getSubscriptionForBundle` with other similar call.

Thanks!

**Comment by javier-gomez-hs** [comment]
@sbrossie 
I've updated the Pull Request with the `changePlan` logic and its test. 

Thanks.

**Comment by sbrossie** [comment]
@javier-gomez-hs The main remark i have is for the `changePlan`: You implemented the logic in the `getDryRunChangePlanEffectiveDate` logic (which works because `entitlement` module calls `dryRun` prior making the change plan). This has the benefit to also implement that logic for dryRun use case, which is nice. A slightly better approach would have been to isolate the logic and call it both from `getDryRunChangePlanEffectiveDate` and `changePlan` (so the two are independent, and if we were to not do dryRun call prior calling `changePlan` that would still work.


If this is easy to fix (< 2 hours of work), in would suggest doing it, if not, this is fine and we can merge it "as is". Let me know whet you think.

**Comment by sbrossie** [comment]
@javier-gomez-hs The main remark i have is for the `changePlan`: You implemented the logic in the `getDryRunChangePlanEffectiveDate` logic (which works because `entitlement` module calls `dryRun` prior making the change plan). This has the benefit to also implement that logic for dryRun use case, which is nice. A slightly better approach would have been to isolate the logic and call it both from `getDryRunChangePlanEffectiveDate` and `changePlan` (so the two are independent, and if we were to not do dryRun call prior calling `changePlan` that would still work.


If this is easy to fix (< 2 hours of work), in would suggest doing it, if not, this is fine and we can merge it "as is". Let me know whet you think.

**Comment by sbrossie** [comment]
@javier-gomez-hs The main remark i have is for the `changePlan`: You implemented the logic in the `getDryRunChangePlanEffectiveDate` logic (which works because `entitlement` module calls `dryRun` prior making the change plan). This has the benefit to also implement that logic for dryRun use case, which is nice. A slightly better approach would have been to isolate the logic and call it both from `getDryRunChangePlanEffectiveDate` and `changePlan` (so the two are independent, and if we were to not do dryRun call prior calling `changePlan` that would still work.


If this is easy to fix (< 2 hours of work), in would suggest doing it, if not, this is fine and we can merge it "as is". Let me know whet you think.

**Comment by javier-gomez-hs** [comment]
@sbrossie 
In order to isolate the logic and call it from both `getDryRunChangePlanEffectiveDate` and `changePlan`, I'll have to do it inside [this method](https://github.com/javier-gomez-hs/killbill/blob/a03d639c1d1c410c3acaa520218b0e814eee266f/entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java#L531-L531) in the `DefaultEntitlement.java` class.
I'll have to inject the CatalogUserApi so I can get the current Catalog and its Plans.
Does that sounds ok for you guys?

Thanks!

**Comment by javier-gomez-hs** [comment]
@sbrossie 
Nevermind my last comment.
I've added the new logic in the [DefaultSubscriptionBaseApiService](https://github.com/javier-gomez-hs/killbill/blob/365ce499f9091bd76576bd4219a1efa948bb42ef/subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBaseApiService.java#L409-L409)

Please review it and let me know if I have to change anything else.

Thanks!

File: subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBaseApiService.java [hunk 2]
```diff
@@ -394,6 +406,16 @@ private void doChangePlan(final DefaultSubscriptionBase subscription,
...
+        if (ProductCategory.ADD_ON.toString().equalsIgnoreCase(newPlan.getProduct().getCategory().toString())) {
+            if (newPlan.getPlansAllowedInBundle() != -1
+                && newPlan.getPlansAllowedInBundle() > 0
+                && countExistingAddOnsWithSamePlanName(dao.getSubscriptions(subscription.getBundleId(), null, internalCallContext), newPlan.getName())
+                   >= newPlan.getPlansAllowedInBundle()) {
+                // the plan can be changed to the new value, because it has reached its limit by bundle
+                throw new SubscriptionBaseApiException(ErrorCode.SUB_CHANGE_AO_MAX_PLAN_ALLOWED_BY_BUNDLE, newPlan.getName());
+            }
+        }
+
```
Comment (sbrossie) [inline subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBaseApiService.java]: We are getting there. One more comment: Is `countExistingAddOnsWithSamePlanName` not the same thing as `countExistingAddOnsWithSamePlanName` in `DefaultSubscriptionInternalApi` ? Can we share the code (maybe a static method that can be used in both classes?)

**Comment by javier-gomez-hs** [comment]
@sbrossie 
I've moved the method to the `AddonUtils`, since both classes are using the same method.
Does that sounds good for you?

Thanks!

**Pull request merged**